File: VIQA/Common/TableLine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using OpenQA.Selenium;
using VIQA.HtmlElements;
using VIQA.HtmlElements.Interfaces;
using VIQA.HtmlElements.SimpleElements;

namespace VIQA.Common
{
    public abstract class TableLine<T> where T : VIElementsSet, IHaveValue
    {
        public virtual List<Cell<T>> this[string name] { get { return null; } }
        public virtual List<Cell<T>> this[int num] { get { return null; } }
        public Func<IWebElement, string[]> GetHeadersFunc;
        public int StartIndex = 1;
        public bool HaveHeaders;
        public ElementIndexType ElementIndex;

        public Table<T> Table;

        protected int? _count;
        public int Count
        {
            set { _count = value; }
            get
            {
                if (_count != null)
                    return (int)_count;
                return Headers != null ? Headers.Length : 0;
            }
        }

        protected string[] _headers;
        public string[] Headers
        {
            set { _headers = value; }
            get
            {
                if (_headers != null)
                    return _headers;
                _headers = Table.DoVIActionResult("Get Rows Headers", () => GetHeadersFunc(Table.GetWebElement()));
                if (_headers == null || !_headers.Any())
                    return default(string[]);
                Count = _headers.Length;
                if (!HaveHeaders)
                    _headers = GetNumList(Count);
                return _headers;
            }
        }

        protected string[] GetNumList(int count, int from = 1)
        {
            var result = new List<string>();
            for (int i = from; i < count + from; i++)
                result.Add(i.ToString());
            return result.ToArray();
        }

        public void Update(TableLine<T> tableLine)
        {
            if (tableLine._count != null)
                Count = tableLine.Count;
            if (tableLine.StartIndex != 1)
                StartIndex = tableLine.StartIndex;
            if (tableLine._headers != null && tableLine._headers.Any())
                Headers = tableLine.Headers;
            if ((tableLine is Columns<T> && !tableLine.HaveHeaders) || (tableLine is Rows<T> && tableLine.HaveHeaders))
                HaveHeaders = tableLine.HaveHeaders;
            if (tableLine.ElementIndex != ElementIndexType.Nums)
                ElementIndex = tableLine.ElementIndex;
            GetHeadersFunc = tableLine.GetHeadersFunc;
        }
    }
    public enum ElementIndexType { Nums, Names }

}

File: Java/src/main/java/ru/viqa/ui_testing/elements/baseClasses/Table/TableLine.java
package ru.viqa.ui_testing.elements.baseClasses.Table;

import ru.viqa.ui_testing.elements.interfaces.IHaveValue;
import ru.viqa.ui_testing.elements.simpleElements.*;

import java.util.ArrayList;
import java.util.List;

import static ru.viqa.ui_testing.common.utils.ReflectionUtils.isClass;

/**
 * Created by 12345 on 25.10.2014.
 */
public abstract class TableLine<T extends IHaveValue> {
    public int startIndex = 1;
    public boolean haveHeader;
    public ElementIndexType elementIndex;

    public Table<T> table;

    protected int count = -1;
    public void setCount(int value) { count = value; }
    public int count() {
        if (count > 0)
            return count;
            return headers != null ? headers.length : 0;
    }

    protected String[] headers;
    public void setHeaders(String[] value) { headers = value; }
    protected abstract String[] getHeadersAction() throws Exception;
    public final String[] headers() throws Exception {
        if (headers != null)
            return headers;
        headers = table.getHeaders(this::getHeadersAction);
        if (headers == null || !(headers.length > 0))
            return null;
        setCount(headers.length);
        if (!haveHeader)
            setHeaders(getNumList(count()));
        return headers;
    }

    protected String[] getNumList(int count) {
        return getNumList(count, 1);
    }
    protected String[] getNumList(int count, int from) {
        List<String> result = new ArrayList<>();
        for (int i = from; i < count + from; i++)
        result.add(i + "");
        return result.toArray(new String[result.size()]);
    }

    public final void update(TableLine<T> tableLine) throws Exception {
        if (tableLine.count > 0)
            setCount(tableLine.count());
        if (tableLine.startIndex != 1)
            startIndex = tableLine.startIndex;
        if (tableLine.headers != null && tableLine.headers.length > 0)
            setHeaders(tableLine.headers());
        if ((isClass(tableLine.getClass(), Columns.class) && !tableLine.haveHeader)
            || (isClass(tableLine.getClass(), Rows.class) && tableLine.haveHeader))
            haveHeader = tableLine.haveHeader;
        if (tableLine.elementIndex != ElementIndexType.Nums)
        elementIndex = tableLine.elementIndex;
    }
}
